util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java

```diff
@@ -0,0 +1,138 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.cafapi.util.flywayinstaller;
+
+import com.github.cafapi.util.flywayinstaller.exceptions.InvalidConnectionStringException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.flywaydb.core.Flyway;
+import org.postgresql.ds.PGSimpleDataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class Migrator
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(Migrator.class);
+    private static final String CONNECTION_URL_REGEX = "^.*\\/\\/.+:\\d+\\/";
+    private static final String CREATE_DATABASE = "CREATE DATABASE \"%s\"";
+    private static final String DOES_DATABASE_EXIST
+        = "SELECT EXISTS ( SELECT datname FROM pg_catalog.pg_database WHERE lower( datname ) = lower( ? ) );";
+
+    private Migrator()
+    {
+    }
+
+    public static void migrate(
+        final String connectionString,
+        final String dbName,
+        final String username,
+        final String password
+    ) throws InvalidConnectionStringException, SQLException
+    {
+        logReceivedArgumentsIfDebug(connectionString, dbName, username, password);
+
+        LOGGER.info("Starting migration ...");
+
+        try {
+            final PGSimpleDataSource dbSource = new PGSimpleDataSource();
+            final String urlAfterConversion = checkAndConvertConnectionUrl(connectionString);
+            dbSource.setUrl(urlAfterConversion);
+            dbSource.setUser(username);
+            dbSource.setPassword(password);
+
+            if (!doesDbExist(dbSource, dbName)) {
+                LOGGER.debug("reset or createDB");
+                resetOrCreateDatabase(dbSource, dbName);
+            }
+            // Update dbSource url to include the database name
+            dbSource.setUrl(urlAfterConversion + dbName);
+
+            LOGGER.info("About to perform DB update.");
+            final Flyway flyway = Flyway.configure()
+                .dataSource(dbSource)
+                .baselineOnMigrate(true)
+                .load();
+            flyway.migrate();
+            LOGGER.info("DB update finished.");
+
+        } catch (final Exception e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+
+        LOGGER.info("Migration completed ...");
+    }
+
+    private static void resetOrCreateDatabase(
+        final PGSimpleDataSource dbSource,
+        final String dbName
+    ) throws SQLException
+    {
+        try (final Connection connection = dbSource.getConnection();
+             final Statement statement = connection.createStatement();) {
+            statement.executeUpdate(String.format(CREATE_DATABASE, dbName));
```

Comment:
Are you sure there isn't a way to use a prepared statement for this?  I feel like there should be.

Reply:
Updated to use prepared statements

Reply:
You're following the letter of the law but not the spirit (i.e. the sql is still being dynamically put together with concatenation - this is what the advice to use prepared statements is meant to avoid).

Reply:
Reading the documentation I don't see a straight forward way of doing this.  The best thing is probably to execute this as a prepared statement:
```
SELECT format('CREATE DATABASE %I', ?)
```

and then to retrieve and execute the result.

Reply:
Nice one, updated.

> Why have you dropped the use of `caf-logging`?
> 
> [30b81aa#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L91-L95](https://github.com/CAFapi/util-flyway-installer/commit/30b81aab736c8d35071481cbac464231c1489ec7#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L91-L95)

@dermot-hardy The initial idea here was to remove the jackson dependency (included in the caf-logging) as it's not being used and as it was bringing conflicts in the job-service. The flyway migrator being used so far solely in the job service, and the job-service already using caf-logging, I thought it would be fine to exclude it from here. But I understand that the utility is supposed to be usable on its own and if so, would expose the logs to possible threats (and also to harmonize log formatting), so I restored it and excluded the jackson dependency from it (317kb).
I also excluded logback-core from it as it was conflicting with the newer version used for compiling: 1.2.10